### PR TITLE
Laushinka/login 6826

### LINKED ISSUE #6826: Customize the Login page for unauthenticated users arriving from an "Open in Gitpod" context
This only applies to the login/signup page presented to unauthenticated users, landing at gitpod.io with a #-prefixed git repo context. Using the repo context we should present those users with a simpler, less-generic Login page to improve the conversion rate for new signups arriving from this source.

- Remove the welcome section on the left
- Show the Gitpod logo and name
- Change "Log in" to "Log in / Sign up"
- Customized message: `Open a cloud based development environment for <repo XXX>`
- Single `Continue` button leading to the git provider matching repo ☝️, usable with the Enter key.

Screenshot 2021-11-21 at 22 06 01

If possible, we would like to implement this in way that will allow us to present either the existing OR the new design, and measure the aggregate signup rates for both versions.   

For further background see [Notion RFC](https://www.notion.so/gitpod/Unqualified-leaky-bucket-d02f1a47aaf44621b5a87e1dcac90de8#6472fae609f040be984bf19ec1a68b1a) (internal)

### PR DESCRIPTION
**Pull request by laushinka**
## Description
Unlogged visitors using a prefix are shown:
1. No welcome section
2. More relevant text
3. A button to continue only with the corresponding Git provider

Screenshot 2021-12-06 at 11 37 35

Screenshot 2021-12-07 at 09 24 29

## Related Issue(s)

Fixes #6826 
Depends on #7081 

## How to test
1. Append a repo URL to `https://laushinka-login-6826.staging.gitpod-dev.com/#`
2. See the changed page

## Release Notes

```release-note
Unlogged visitors using a prefix will be shown a more direct login page.
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#6826*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/Login.tsx
Comment (jankeromnes) [inline components/dashboard/src/Login.tsx:196]: **Thought**: This is very much personal preference, but I personally prefer keeping all HTML markup together, because I think it helps quickly understanding how the page is built, and facilitates making structural changes in the future. I find that splitting the `render()` function into smaller `renderXYZ()` functions breaks apart the markup too much, and makes it harder to rework it later.

E.g. compare these pseudo examples:

```jsx
    return <>
        <Header title="my page" />
        <p>Manage your elements</p>
        {elements.length > 0
            ? <>
                <h2>My Elements</h2>
                <List elements={elements} />
            </>
            : <>
                <h2>No Elements Yet!</h2>
                <Button onClick={createElement()}>New Element</Button>
            </>}
    </>;
```

with:

```jsx
    const renderSubtitle = () => {
        return <p>Manage your elements</p>;
    }

    const renderEmptyState = () => {
        return <>
            <h2>No Elements Yet!</h2>
            <Button onClick={createElement()}>New Element</Button>
        </>;
    }

    const renderElementsList = (elements) => {
        return <>
            <h2>My Elements</h2>
            <List elements={elements} />
        </>;
    }

    const renderElementsPage = (elements) => {
        if (elements.length === 0) {
            return renderElementsList(elements);
        }
        return renderEmptyState();
    }

    return <>
        <Header title="my page" />
        {renderSubtitle()}
        {renderElementsPage()}
    </>;
```

Okay, maybe my example is a little bit contrived, but the second example is so hard to understand and navigate 🙈 you need to jump back and forth across the code, and the HTML markup is split all over the place in random order.

I'd find the first example much easier to rework (e.g. if you need to restyle some part, or move the markup under another `<div>` layer, or restructure the page a bit by making more things conditional, etc).

But again, this is totally personal, so please feel free to ignore if you do find sub-render functions better! 😁

File: components/dashboard/src/App.test.ts
Comment (jankeromnes) [inline components/dashboard/src/App.test.ts:15]: Many thanks for adding tests! 💯 👏

I'm just slightly confused as to what problem this test tries to address. In which situation can your `window.location` have both a `href` with `#` _and_ a `hash`? 🤔 Is this related to a bug we've seen? (Maybe a small comment could help clarify this for future readers 😇)

**Review by jankeromnes — COMMENTED** [review]
Many thanks for improving the login screen UX!

Unfortunately, there still seems to be some uncertainty around what exactly the page should look like in https://github.com/gitpod-io/gitpod/issues/6826, but I guess this PR can be adjusted when the final design is agreed upon.

(I have a slight suspicion that we might move to something like https://github.com/gitpod-io/gitpod/issues/6826#issuecomment-975469912 but with a dropdown element to the `Continue with GitHub` button as shown in `PROPOSAL` in https://github.com/gitpod-io/gitpod/issues/6826#issuecomment-985039468 -- but let's see).

I gave this a try, and it seems to work as intended in various scenarios. 👍

Except, however, that it always shows the welcome screen for me (while the PR description says "1. No welcome section"). Maybe that's unintended?

| No repo | With GitHub repo | With GitLab repo |
| --- | --- | --- |
| Screenshot 2021-12-03 at 09 35 19 | Screenshot 2021-12-03 at 09 35 43 | Screenshot 2021-12-03 at 09 36 08 |

Also left one minor thoughts in the code, but please let me know if you'd like a more thorough code review at this point -- happy to do that! 😊

**Comment by laushinka** [comment]
> Except, however, that it always shows the welcome screen for me (while the PR description says "1. No welcome section"). Maybe that's unintended?

@jankeromnes Oh this shouldn't have happened. I will check, thanks!

> please let me know if you'd like a more thorough code review at this point

I'm always happy with that! We can also do it synchronously if that works for you? 🤗

File: components/dashboard/src/App.test.ts
Comment (laushinka) [reply components/dashboard/src/App.test.ts:15]: Lol I completely missed this. Initially I wanted to add this simple test because my Regex skills is pretty non-existent (and we have a bunch in App.tsx), so I wanted to continue playing around with tests but ended up forgetting about it and focusing on the feature. Will fix this, and if we do a sync review maybe we can try more tests together.

**Comment by jankeromnes** [comment]
> > please let me know if you'd like a more thorough code review at this point
> 
> I'm always happy with that! We can also do it synchronously if that works for you? 🤗

Sure! More than happy to sync on this -- how about this afternoon, e.g. 14:00 or 15:00? (Or please feel free to pick any time you like [here](https://calendly.com/jan-keromnes/30min) 😁)

I often feel like we should do more pair programming at Gitpod, but somehow there is always a bit of friction to set it up, so I end up never doing it. 🙈 Thanks for providing this opportunity! 🙏

**Comment by laushinka** [comment]
> I often feel like we should do more pair programming at Gitpod, but somehow there is always a bit of friction to set it up, so I end up never doing it. 🙈 Thanks for providing this opportunity! 🙏

💯💯💯💯💯💯

File: components/dashboard/src/Login.tsx
Comment (jankeromnes) [reply components/dashboard/src/Login.tsx:196]: Good point, thank you for sharing your view on this. I know others in the team prefer split render functions as well (hi @AlexTugarev 👋) so I might be in the minority here. I'll make more efforts to adapt. 😁

**Review by Foge9627 — APPROVED** [review]

**Comment by roboquat** [comment]
@Foge9627: changing LGTM is restricted to collaborators

<details>

In response to [this](https://github.com/gitpod-io/gitpod/pull/7046#pullrequestreview-822495063):

>


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7046#pullrequestreview-822495063" title="Approved">Foge9627</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#6826*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/App.test.ts
Comment (laushinka) [reply components/dashboard/src/App.test.ts:15]: Oh I misread the question, I thought you meant my `href` was wrong (which it is) but `href` and `hash` can both be there. The `href` is basically the same as `window.location`, and the `hash` is only the part starting from the `#`.

**Comment by laushinka** [comment]
> Except, however, that it always shows the welcome screen for me (while the PR description says "1. No welcome section"). Maybe that's unintended?

@jankeromnes Push fixed! We can still go through the code later in the call :)

**Comment by laushinka** [comment]
Build finally green again after werft issue fixed. Ready for another code review if you'd like @jankeromnes. (I don't mean this weekend of course 🙈)

**Comment by jldec** [comment]
Thanks @laushinka - nice and clean :)

I'm missing the Gitpod name on the page somewhere.

The "Log in / Sign up" should not be bigger than the line below it, which carries an equally important messge.
Screenshot 2021-12-06 at 00 24 35

I got into trouble trying to login with GitLab (ended up on another login page with 3 buttons) - I would suggest giving the GitLab flow a little more testing.

**Comment by svenefftinge** [comment]
Looks like the version from the first comment was implemented and not the one that I understood [was agreed on](https://github.com/gitpod-io/gitpod/issues/6826#issuecomment-975469912). Is that intentional?

**Comment by jankeromnes** [comment]
> Build finally green again after werft issue fixed.

Hooray! 🙌

> Ready for another code review if you'd like @jankeromnes. (I don't mean this weekend of course 🙈)

Thanks Laurie! 🙏😊 Sure, I'm happy to do another code review (could you please re-request review in the top right of this PR? This helps me differentiate PRs between "work in progress" and "blocked until reviewed" 😇). However, I also agree with Sven: Could you first please adjust the design to [this version](https://github.com/gitpod-io/gitpod/issues/6826#issuecomment-975469912)? I agree the issue isn't very clear (so many design variations 🙈) but I think [that one](https://github.com/gitpod-io/gitpod/issues/6826#issuecomment-975469912) is the current best compromise.

**Comment by laushinka** [comment]
> Looks like the version from the first comment was implemented and not the one that I understood [was agreed on](https://github.com/gitpod-io/gitpod/issues/6826#issuecomment-975469912). Is that intentional?

@svenefftinge I can do this version - it did get confusing inside the issue, but I will take the comments in this PR as final, and specify that in the issue as well. cc @jldec to make sure we're aligned.

**Comment by laushinka** [comment]
> I got into trouble trying to login with GitLab (ended up on another login page with 3 buttons) - I would suggest giving the GitLab flow a little more testing.

@jldec Would you mind taking screenshots or recording it, or showing me sync later? 🙏🏽

**Comment by laushinka** [comment]
> could you please re-request review in the top right of this PR?

@jankeromnes TIL about the button!

**Comment by svenefftinge** [comment]
> > Looks like the version from the first comment was implemented and not the one that I understood [was agreed on](https://github.com/gitpod-io/gitpod/issues/6826#issuecomment-975469912). Is that intentional?
> 
> @svenefftinge I can do this version - it did get confusing inside the issue, but I will take the comments in this PR as final, and specify that in the issue as well. cc @jldec to make sure we're aligned.

No worries, I'm also confused 😅 Let's wait for @jldec to clarify.

**Comment by jldec** [comment]
Confirming that we should implement the design as agreed after discussion in #6826 (copied for reference below)

 142861310-c32ae967-93ed-4055-995e-d03ff7ffd21e

**Comment by laushinka** [comment]
Ready for review @jankeromnes @gtsiolis @jldec.
@jldec I tested it with GitLab, but let me know if you still have issues with it - you can also show me in a quick call. 🙏🏽

**Comment by jankeromnes** [comment]
Many thanks for adjusting this @laushinka! 🙏 Taking another look now. 👀

File: components/dashboard/src/App.test.ts
Comment (jankeromnes) [inline components/dashboard/src/App.test.ts:17]: **Optional**: Maybe we can make the example URL look more like a contextURL:

```suggestion
            hash: '#https://example.org/user/repo'
        }
    });

    expect(getURLHash()).toBe('https://example.org/user/repo');
```

But this is entirely optional of course. 👍

File: components/dashboard/src/Login.tsx
Comment (jankeromnes) [inline components/dashboard/src/Login.tsx:57]: The additional `new URL()` outside the try/catch can throw again. I'd suggest putting them in common inside the try/catch:

```suggestion
    let hostFromContext: string | undefined;
    let repoPathname: string | undefined;
    try {
        const url = new URL(urlHash);
        hostFromContext = url.host;
        repoPathname = url.pathname;
    } catch (error) {
        // Hash is not a valid URL
    }
```

File: components/dashboard/src/Login.tsx
Comment (jankeromnes) [inline components/dashboard/src/Login.tsx:163]: **Optional**: I think it's more common to mention repositories as `gitpod-io/gitpod` rather than `/gitpod-io/gitpod`, so I'd slightly deviate from the specification here and remove the first slash:

```suggestion
                                    <h2 className="text-xl">for the repository {repoPathname.slice(1)}</h2>
```

File: components/dashboard/src/Login.tsx
Comment (jankeromnes) [inline components/dashboard/src/Login.tsx:175]: Maybe give this a _tiny_ bit more space when it's the only button:

```suggestion
                                    <button key={"button" + providerFromContext.host} className="mt-2 btn-login flex-none w-56 h-10 p-0 inline-flex" onClick={() => openLogin(providerFromContext.host)}>
```

**Review by jankeromnes — COMMENTED** [review]
This is getting super nice! ✨

Added a few comments, but it's mostly for dark theme support.

Other than that, it's an excellent first iteration 🎸 (there will likely be more adjustments to this page in the future when we learn more from user feedback 🔧 but I'd be happy to see this merged as is)

File: components/dashboard/src/Login.tsx
Comment (laushinka) [reply components/dashboard/src/Login.tsx:57]: Of course! Thanks, this should've been obvious to me 🙈

File: components/dashboard/src/Login.tsx
Comment (laushinka) [reply components/dashboard/src/Login.tsx:163]: cc @jldec?

Comment (laushinka) [reply components/dashboard/src/Login.tsx:163]: I like having the slash, but I don't know why and can't explain it 🤣 😅 So I'll go with what @jldec approves of.

File: components/dashboard/src/Login.tsx
Comment (laushinka) [reply components/dashboard/src/Login.tsx:156]: Thank you for reminding me about the dark theme!

File: components/dashboard/src/Login.tsx
Comment (jldec) [reply components/dashboard/src/Login.tsx:163]: Let's drop the leading /

**Comment by laushinka** [comment]
Updated for dark theme - please check the screenshot whether that looks correct as I didn't adapt the repo text.

**Review by Foge9627 — APPROVED** [review]

**Comment by roboquat** [comment]
@Foge9627: changing LGTM is restricted to collaborators

<details>

In response to [this](https://github.com/gitpod-io/gitpod/pull/7046#pullrequestreview-824957264):

>


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7046#pullrequestreview-824957264" title="Approved">Foge9627</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **laushinka** after the PR has been reviewed.

Associated issue: *#6826*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/Login.tsx
Comment (geropl) [inline components/dashboard/src/Login.tsx:66]:
```
const showWelcome = Experiment.has("login-6826") ?
     (!hasLoggedInBefore() && !hasVisitedMarketingWebsiteBefore() && !urlHash.startsWith("https://"))
   : (!hasLoggedInBefore() && !hasVisitedMarketingWebsiteBefore())
;
```

File: components/dashboard/src/Login.tsx
Comment (geropl) [inline components/dashboard/src/Login.tsx:75]:
```
if (Experiment.has("login-6826")) {
   useEffect(...);
}
```

**Comment by laushinka** [comment]
Will plug in the experiments code once [this](https://github.com/gitpod-io/gitpod/pull/7081) is merged.

**Comment by laushinka** [comment]
@geropl Please review the experiments code 🙏🏽

**Comment by jankeromnes** [comment]
Thanks again @laushinka! Taking another look now/soon 😇

File: components/dashboard/src/experiments.ts
Comment (geropl) [inline components/dashboard/src/experiments.ts:31]: Some documentation incl. a link to this PR would be nice!

File: components/dashboard/src/Login.tsx
Comment (geropl) [inline components/dashboard/src/Login.tsx:51]: We should use the string directly whenever necessary, no need to store in a `const` here. The compiler will tell us if it's mistyped/outdated. :+1:

File: components/dashboard/src/Login.tsx
Comment (geropl) [inline components/dashboard/src/Login.tsx:71]: :+1:

File: components/dashboard/src/experiments.ts
Comment (jankeromnes) [reply components/dashboard/src/experiments.ts:31]: Also, side question: Can you manually "force" A or B to test code-dev deployments?

File: components/dashboard/src/Login.tsx
Comment (geropl) [inline components/dashboard/src/Login.tsx:82]: :+1:

Comment (geropl) [reply components/dashboard/src/Login.tsx:82]: @laushinka  My bad, turns our React does not like this: build failed with `Line 83:9:  React Hook "useEffect" is called conditionally. React Hooks must be called in the exact same order in every component render  react-hooks/rules-of-hooks`.

Then the `if(Experiment.hat...` goes into the `useEffect`, on top-level. :+1:

File: components/dashboard/src/experiments.ts
Comment (geropl) [reply components/dashboard/src/experiments.ts:31]: > Can you manually "force" A or B to test code-dev deployments?

Sure, you have to manipulate `localStorage` in your web dev tools. They are configured to be stable atm, so will happily accept your changes.

**Update**: just recognized that this is not true, actually! And found a bug in the impl. Will fix after lunch. :upside_down_face:

File: components/dashboard/src/Login.tsx
Comment (jankeromnes) [inline components/dashboard/src/Login.tsx:82]: React doesn't seem to like the conditional `useEffect`.

Maybe always run the `useEffect`, but do `if (!Experiment.has(experiment)) { return; }` at the top?

File: components/dashboard/src/Login.tsx
Comment (jankeromnes) [inline components/dashboard/src/Login.tsx:172]: Oops, this doesn't look right:
- In light theme, we conditionally switch between Gitpod icon and Gitpod full logo
- But in dark theme, we always use Gitpod full logo, regardless of whether we have `providerFromContext`?

Could be fixed like so:

```suggestion
                            <img src={providerFromContext ? gitpod : gitpodIcon} className="h-14 mx-auto block dark:hidden" alt="Gitpod's logo" />
                            <img src={providerFromContext ? gitpodDark : gitpodIcon} className="h-14 hidden mx-auto dark:block" alt="Gitpod dark theme logo" />
```

File: components/dashboard/src/Login.tsx
Comment (laushinka) [reply components/dashboard/src/Login.tsx:82]: Yes just saw this as well

File: components/dashboard/src/Login.tsx
Comment (laushinka) [reply components/dashboard/src/Login.tsx:172]: Thanks 🤦🏻‍♀️

File: components/dashboard/src/Login.tsx
Comment (laushinka) [reply components/dashboard/src/Login.tsx:82]: Yes just realized this too

File: components/dashboard/src/Login.tsx
Comment (laushinka) [reply components/dashboard/src/Login.tsx:172]: I clicked `commit suggestion` but it didn't seem to do it..

**Comment by laushinka** [comment]
@geropl I have my local storage set but not seeing the changes. 
Screenshot 2021-12-07 at 13 03 45
..

File: components/dashboard/src/experiments.ts
Comment (geropl) [reply components/dashboard/src/experiments.ts]: PR: https://github.com/gitpod-io/gitpod/pull/7103

**Comment by geropl** [comment]
@laushinka Sorry for the additional confusion - and thx for fixing. :pray: Here's a follow-up PR to fix other issues as well: https://github.com/gitpod-io/gitpod/pull/7103

**Review by geropl — APPROVED** [review]
Cannot test - "max number of users for licence" reached :shrug: 

But tested before, and code lgtm, so: :+1:

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 26cdb08ca8eb7574a013df588f0031e3827e8a76</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7046#pullrequestreview-824957264" title="Approved">Foge9627</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7046#pullrequestreview-825229585" title="Approved">geropl</a>*

Associated issue: *#6826*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [geropl]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by jldec** [comment]
See followups #7188 and #7138

**Pull request merged**
